server/spamTag/communities.ts

```diff
@@ -0,0 +1,83 @@
+import { Op } from 'sequelize';
+
+import * as types from 'types';
+import { Community, SpamTag } from 'server/models';
+
+const orderableFields: Record<types.SpamCommunityQueryOrderingField, any[]> = {
+	'community-created-at': ['createdAt'],
+	'spam-status-updated-at': [{ model: SpamTag, as: 'spamTag' }, 'statusUpdatedAt'],
+	'spam-score': [{ model: SpamTag, as: 'spamTag' }, 'spamScore'],
+};
+
+const getWhereQueryPartForUrl = (searchTerm: string) => {
+	let url: URL;
+	try {
+		url = new URL(searchTerm);
+	} catch (err: unknown) {
+		return [];
+	}
+	if (url.hostname.endsWith('.pubpub.org')) {
+		const subdomain = url.hostname.replace('.pubpub.org', '');
+		return [{ subdomain }];
+	}
+	return [{ domain: url.hostname }];
+};
+
+const getWhereQueryPartForFreeformEntry = (searchTerm: string) => {
+	const normalizedQuery = `%${searchTerm.trim()}%`;
+	return {
+		[Op.or]: [
+			{ title: { [Op.iLike]: normalizedQuery } },
+			{ subdomain: { [Op.iLike]: normalizedQuery } },
+			{ domain: { [Op.iLike]: normalizedQuery } },
+		],
+	};
+};
+
+const getCommunityWhereQuery = (searchTerm: undefined | string) => {
+	if (searchTerm) {
+		return {
+			where: {
+				[Op.or]: [
+					getWhereQueryPartForFreeformEntry(searchTerm),
+					getWhereQueryPartForUrl(searchTerm),
+				],
```

Comment:
why use two queries when one always a subdomain?

Reply:
Good question — the first one is comparing the `searchTerm` against directly against e.g. the `subdomain` field, so if  you type in `kfg-notes` then it will find that Community. The second is checking whether the `searchTerm` is a valid URL, and if so, it's extracting parts of that URL and checking _those_ against e.g. `subdomain`. So that'll match if you paste in `https://kfg-notes.pubpub.org` (or since we're also checking the `domain` field) `https://notes.knowledgefutures.org`.